examples/nas/multi-trial/mnist/search.py

```diff
@@ -131,15 +131,16 @@ def evaluate_model(model_cls):
     exp_config = RetiariiExeConfig('local')
     exp_config.experiment_name = 'mnist_search'
     exp_config.trial_concurrency = 2
-    exp_config.max_trial_number = 20
+    exp_config.max_trial_number = 4
     exp_config.training_service.use_active_gpu = False
     export_formatter = 'dict'
 
     # uncomment this for graph-based execution engine
     # exp_config.execution_engine = 'base'
     # export_formatter = 'code'
 
-    exp.run(exp_config, 8080)
+    exp.run(exp_config, 8090)
     print('Final model:')
     for model_code in exp.export_top_models(formatter=export_formatter):
         print(model_code)
+    exp.stop()
```

Comment:
yes. Another option is putting this behavior in the new NasExperiment interface while keeping the previous behavior in RetiariiExperiment. Explicitly stopping an experiment is a good style.

Reply:
If I understand correctly, `run = start + stop`?

Reply:
Seems I haven't made it clear.
In v2.7 HPO experiment, `stop()` is recommended but not enforced. `run()` waits for completion but does not invoke `stop()`.

Consider 2 use cases:

###  Case A

```python
exp = Experiment(...)
exp.run()
# end of code
```

The experiment will stop when done. This is (effectively) the same as v2.6.
At the moment when `exp.run()` returns, the experiment is still serving web portal. But right after that, Python finds there is no more code to run and starts exiting. Then Python triggers `atexit` hooks, which will invoke `exp.stop()` to stop NNI manager.

### Case B

```python
exp = Experiment(...)
exp.run()
input()
```

The experiment will serve web portal forever until the user presses enter. This is different from v2.6.
In v2.6 the experiment stops serving web portal when `exp.run()` returns. So the user could not view web portal, or use REST API functions to query experiment result, even though the Python program is still running.

Reply:
got it!

Reply:
it is needed, otherwise the python script won't exit. because different from hpo experiment, retiarii experiment starts advisor in a thread which waits for nnimanager to be stopped

Reply:
Why atextit can't register a stop?

Reply:
stop already registered, but the main script does not exit due to a running sub-thread (advisor)

Reply:
atexit triggers stop. Stop kills nnimanager, advisor detects the termination of nnimanager and exits.